upup/pkg/fi/cloudup/urls.go

```diff
@@ -17,74 +17,162 @@ limitations under the License.
 package cloudup
 
 import (
+	"net/url"
 	"os"
 	"strings"
 
+	"path"
+
 	"github.com/golang/glog"
 	"k8s.io/kops"
+	"k8s.io/kops/pkg/assets"
+	"k8s.io/kops/util/pkg/hashing"
 )
 
-// baseUrl caches the BaseUrl value
-var baseUrl string
+var kopsUrl = "https://kubeupv2.s3.amazonaws.com/kops/" + kops.Version + "/"
+var baseUrl *url.URL
 
 // BaseUrl returns the base url for the distribution of kops - in particular for nodeup & docker images
-func BaseUrl() string {
-	if baseUrl != "" {
+func BaseUrl() (*url.URL, error) {
```

Comment:
Rather than using a string, we are now using a URL.  This provides more error handling and also has helped me debug.  No of the functionality has changed, but now we are using `path` to build URLs, rather than catenating strings.